File: WildWatchAPI/Models/Sighting.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Driver;
using MongoDB.Driver.GeoJsonObjectModel;
using System.Text.Json.Serialization;

namespace WildWatchAPI.Models
{
    public class SightingSummaryUser
    {
        public MongoDBRef Id { get; set; } = null!;
        public DateTime SightingTime { get; set; }
        public GeoJson2DGeographicCoordinates Location { get; set; } = null!;
        public SpeciesSummary Species { get; set; }=null!;
    }

    public class SightingSummarySpecies
    {
        public MongoDBRef Id { get; set; } = null!;
        public DateTime SightingTime { get; set; }
        public GeoJson2DGeographicCoordinates Location { get; set; } = null!;

        public UserSummary Sighter { get; set; } = null!;
    }

    public class SightingSummaryHabitat
    {
        public MongoDBRef Id { get; set; } = null!;
        public DateTime SightingTime { get; set; }
        public GeoJson2DGeographicCoordinates Location { get; set; } = null!;
        public SpeciesSummary Species { get; set; } = null!;

        public UserSummary Sighter { get; set; } = null!;
    }

    public class Sighting
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string? Id { get; set; }
        public DateTime SightingTime { get; set; }


        public GeoJson2DGeographicCoordinates Location { get; set; } = null!;
        [JsonIgnore]
        public UserSummary Sighter { get; set; } = null!;

        [JsonIgnore]
        public SpeciesSummary Species { get; set; } = null!;
        public string? ImageUrl { get; set; }

        [BsonExtraElements]
        public BsonDocument? AdditionalInformation { get; set; }

    }
}

File: Frontend/wild-watch-frontend/src/models/Sighting.ts
import {SpeciesSummary} from "./Species"
import { UserSummary } from "./User"
import { LocationI } from "../types/location"


export class SightingSummaryUser{
    id:string
    sightingTime:Date
    location:LocationI
    species:SpeciesSummary
    imageUrl:string
    constructor(
        id:string,
        sightingTime:Date,
        location:LocationI,
        species:SpeciesSummary,
        imageUrl:string
    ){
        this.id=id;
        this.sightingTime=sightingTime;
        this.location=location;
        this.species=species;
        this.imageUrl=imageUrl;
    }
}

export class SightingSummarySpecies{
    id:string
    sightingTime:Date
    location:LocationI
    sighter:UserSummary
    imageUrl:string

    constructor(
        id:string,
        sightingTime:Date,
        location:LocationI,
        sighter:UserSummary,
        imageUrl:string
    ){
        this.id=id;
        this.sightingTime=sightingTime;
        this.location=location;
        this.sighter=sighter;
        this.imageUrl=imageUrl;
    }
}

export class SightingSummaryHabitat{
    id:string
    sightingTime:Date
    location:LocationI
    sighter:UserSummary
    species:SpeciesSummary
    imageUrl:string

    constructor(
        id:string,
        sightingTime:Date,
        location:LocationI,
        sighter:UserSummary,
        species:SpeciesSummary,
        imageUrl:string
    ){
        this.id=id;
        this.sightingTime=sightingTime;
        this.location=location;
        this.sighter=sighter;
        this.species=species;
        this.imageUrl=imageUrl;
    }
}

export class Sighting{
    id:string
    sightingTime:Date
    location:LocationI
    sighter:UserSummary
    species:SpeciesSummary
    imageUrl:string
    constructor(
        id:string,
        sightingTime:Date,
        location:LocationI,
        sighter:UserSummary,
        species:SpeciesSummary,
        imageUrl:string
    ){
        this.id=id;
        this.sightingTime=sightingTime;
        this.location=location;
        this.sighter=sighter;
        this.species=species;
        this.imageUrl=imageUrl;
    }
}

// public class SightingDto
// {
//     public DateTime SightingTime { get; set; }
//     public GeoJson2DGeographicCoordinates Location { get; set; } = null!;
//     public string? PhotoUrl { get; set; }

//     public string SighterId { get; set; } = null!;

//     public Klasa SpeciesClass { get; set; }
//     public string CommonName { get; set; } = null!;

//     public string ScientificName { get; set; } = null!;

//     public string? ImageUrl { get; set; }
//     public string Description { get; set; } = null!;

//     public string ConservationStatus { get; set; }= null!;

//     public string? Comment { get; set; }

// }
export class GeoJson2DGeographicCoordinates{
    longitude: number
    latitude: number

    constructor(lat: number, long: number){
        this.longitude = long
        this.latitude = lat
    }
}

export class SightingDto {
    sightingTime: Date
    location: GeoJson2DGeographicCoordinates
    photoUrl: string
    sighterId: string
    speciesClass: number
    commonName: string
    scientificName: string
    imageUrl: string
    description: string
    conservationStatus: string
    comment: string

    constructor(
        sightingTime: Date,
        location: GeoJson2DGeographicCoordinates,
        photoUrl: string,
        sighterId: string,
        speciesClass: number,
        commonName: string,
        scientificName: string,
        imageUrl: string,
        description: string,
        conservationStatus: string,
        comment: string
    ) {
        this.sightingTime = sightingTime;
        this.location = location;
        this.photoUrl = photoUrl;
        this.sighterId = sighterId;
        this.speciesClass = speciesClass;
        this.commonName = commonName;
        this.scientificName = scientificName;
        this.imageUrl = imageUrl;
        this.description = description;
        this.conservationStatus = conservationStatus;
        this.comment = comment;
    }
}
